### DISCUSSION TITLE
Update Environment

### DISCUSSION THREAD

**Discussion by prateekbhustali** [AI-for-Infrastructure-Management/imp-act#52] [General]
## Initial damage condition

| p(s) 	|  p(s5)  	|  p(s4)  	|  p(s3)  	|  p(s2)  	|  p(s1)  	|
|------	|:-------:	|:-------:	|:-------:	|:-------:	|:-------:	|
|      	| state 5 	| state 4 	| state 3 	| state 2 	| state 1 	|
| ME   	|    0.55 	|   0.150 	|   0.150 	|   0.150 	|       0 	|
| DK   	|    0.72 	|   0.093 	|   0.093 	|   0.093 	|       0 	|
| BE   	|     0.6 	|   0.133 	|   0.133 	|   0.133 	|       0 	|

## Transition model

We now have five actions and  transition tables for all is in the [Excel](https://docs.google.com/spreadsheets/d/1pwiwBTwk5ZjLTIyfJLdtUu3x0jBikVdONnLs2bMqCxg/edit?usp=sharing)

- Do-nothing

- High-fidelity inspection

- Minor-repair: (needs 28.125 days, finishes within 1 timestep)

- Major-repair: (needs 106.25 days, finishes within 1 timestep)

- Reconstruction: (needs 365days, finishes in 1 timestep)

(- Forced repair: If component is in the final damage state, reconstruction is forced. Additionally, forced reconstruction has higher costs.)

## Observation model

We get low-fidelity inspection at every timestep. However, if agent takes a high-fidelity inspection action, it overrides the low-fidelity inspection.

#### Low-fidelity inspection

|         	| obs 5 	| obs 4 	| obs 3 	| obs 2 	| obs 1 	|
|---------	|:-----:	|:-----:	|:-----:	|:-----:	|:-----:	|
| state 5 	|  0.8  	|  0.2  	|       	|       	|       	|
| state 4 	|  0.2  	|  0.6  	|  0.2  	|       	|       	|
| state 3 	|       	|  0.2  	|  0.6  	|  0.2  	|       	|
| state 2 	|       	|       	|  0.2  	|  0.6  	|  0.2  	|
| state 1 	|       	|       	|       	|  0.2  	|  0.80 	|

#### High-fidelity inspection
|         	| obs 5 	| obs 4 	| obs 3 	| obs 2 	| obs 1 	|
|---------	|:-----:	|:-----:	|:-----:	|:-----:	|:-----:	|
| state 5 	|  0.95 	|  0.05 	|       	|       	|       	|
| state 4 	|  0.05 	|  0.90 	|  0.05 	|       	|       	|
| state 3 	|       	|  0.05 	|  0.95 	|  0.05 	|       	|
| state 2 	|       	|       	|  0.05 	|  0.9  	|  0.05 	|
| state 1 	|       	|       	|       	|  0.05 	|  0.95 	|

- Do-nothing --> low-fidelity inspection 

- High-fidelity inspection

- Minor-repair --> low-fidelity inspection 

- Major-repair --> low-fidelity inspection 

- Reconstruction --> no inspection

![Image](https://github.com/orgs/AI-for-Infrastructure-Management/projects/1/assets/61668356/5b8feb0c-f0df-4548-8e62-c2b08a69ac97)

## Reward model

|          	|  state 5 	|  state 4 	|  state 3 	|  state 2 	|  state 1 	|
|----------	|:--------:	|:--------:	|:--------:	|:--------:	|:--------:	|
| action 1 	| 3.00E+03 	| 3.00E+03 	| 3.00E+03 	| 3.00E+03 	| 3.00E+03 	|
| action 2 	| 1.00E+04 	| 1.00E+04 	| 1.00E+04 	| 1.00E+04 	| 1.00E+04 	|
| action 3 	| 2.00E+06 	| 2.00E+06 	| 2.00E+06 	| 2.00E+06 	| 2.00E+06 	|
| action 4 	| 7.50E+06 	| 7.50E+06 	| 7.50E+06 	| 7.50E+06 	| 7.50E+06 	|
| action 5 	| 3.50E+07 	| 3.50E+07 	| 3.50E+07 	| 3.50E+07 	| 4.20E+07 	|

## Objective

$\sum (maintenance) + wage  \times  \Delta_{ttt}$

wage: trucker hourly wage
$\Delta_{ttt}$: increase in total travel time in the network

## Capacity model

f: fraction of lanes closed
d: days for which it is closed
c'/c: updated ratio to be filled in the table.

|              	| lanes-affected 	|   f  	|    d   	|  c'/c  	|
|--------------	|----------------	|:----:	|:------:	|:------:	|
| Do-nothing   	|        0       	|   0  	|    0   	| 1.0000 	|
| Inspect      	|        0       	|   0  	|    0   	| 1.0000 	|
| Minor-repair 	|        1       	| 0.75 	| 28.125 	| 0.9807 	|
| Major-repair 	|        1       	| 0.75 	| 106.25 	| 0.9272 	|
| Replacement  	|        2       	|  0.5 	|   365  	| 0.5000 	|

## Base Capacity

To calclulate base capcity of the segments. [FHWA formula- page 4](https://www.fhwa.dot.gov/policyinformation/pubs/pl18003/hpms_cap.pdf)

|  Capacity  	|             	|               	|
|:----------:	|:-----------:	|:-------------:	|
| Free Speed 	|          80 	| kmph          	|
|            	|          50 	| mph           	|
| %HV        	|          50 	|               	|
|            	|             	|               	|
|            	| 1466.666667 	| vehicles/hour 	|
|            	|    1.28E+07 	| vehicles/year 	|

% HV is percentage heavy vehicles.

## Base travel time model

|              	| new speed 	| # days 	|  b'/b  	|
|--------------	|:---------:	|:------:	|:------:	|
| Do-nothing   	|     80    	|    0   	| 1.0000 	|
| Inspect      	|     80    	|    0   	| 1.0000 	|
| Minor-repair 	|     60    	| 28.125 	| 1.0196 	|
| Major-repair 	|     60    	| 106.25 	| 1.0785 	|
| Replacement  	|     60    	|   365  	| 1.3333 	|

All formulas are available in the Excel in case you want to change any values.

**Comment by DanielHettegger** [comment]
Thanks for the hard work and the great summary, Prateek! 

Here are some of my thoughts on this: 

- Regarding the states, it seems that we are leaving state 0 as best in the code but are referring to it as state 5 in the docs, correct?
- I would really prefer if the replace action would not take more than one time step (We can double the cost and half the time as discussed). Because I feel like this makes the problem less nice to work with from a participant's view. But if we want to keep it, we need to have it observable to the agent in some way that the next action will be overridden.
- I feel like the low-fidelity inspections are already very informative; I wonder if the agent will ever need the high-fidelity ones. On the other hand, the high-fidelity inspections are almost perfect information about the state, which might reduce the POMDP aspect of the problem somewhat. 
- In the high-fidelity table, some rows do not sum to 1, but I think that is a copy-paste error.
- Rewards Q1: I would say if we halve the segment length, the speed and costs would double; therefore, it seems reasonable that there is a linear correlation between the two.
-Rewards Q2: Normalizing rewards should be done by participants if needed IMO.
- Objective: I think we could include a bit more than just wage of the drivers, the cost of delayed goods, missed opportunity cost of next shipments, and worse utilization of the truck are all more costly if it takes more time to deliver I would say, but I am not sure how to arrive at a concrete number with this. 
- 80 km/h base travel speed seems a bit slow. Did we not have 100 in mind?
- The capacity calculation uses 100% utilization for 24h/day for 356 days. Seems a bit much, so we could probably half this I would say.

Let me know what all of you think about this. I think in general, we are arriving at a very nice model detail to feasibility trade-off.

**Reply by prateekbhustali** [reply]
Thanks for the detailed comments!

(At this point, it seems modeling the environment is just as challenging (if not more) than solving it)

**Comment by prateekbhustali** [comment]
> Regarding the states, it seems that we are leaving state 0 as best in the code but are referring to it as state 5 in the docs, correct?

yes, that's correct. I decided to stick to our convention to avoid confusion, and minimize changes. 

> I would really prefer if the replace action would not take more than one time step (We can double the cost and half the time as discussed). Because I feel like this makes the problem less nice to work with from a participant's view. But if we want to keep it, we need to have it observable to the agent in some way that the next action will be overridden.

I agree. From an implementation standpoint (particularly wrt the JAX implementation), finishing reconstruction within 1 timestep is best. Reconstruction requires 42 [days/lane-mile] x (10 [km]/1.6) [miles] x 4 [lanes] = 1050 days. To finish within 365 days, reconstruction needs to happen at 1050/365 ~ 3x speed. Building a 4-lane road of 10km within 1 year seems feasible to me.

> I feel like the low-fidelity inspections are already very informative; I wonder if the agent will ever need the high-fidelity ones. On the other hand, the high-fidelity inspections are almost perfect information about the state, which might reduce the POMDP aspect of the problem somewhat.

If low-fidelity inspections are good-enough, agent(s) will only rely on those. I think this is fine.

> In the high-fidelity table, some rows do not sum to 1, but I think that is a copy-paste error.

indeed. In row 2, '0.95' should be '0.9'. I have added a `_check_model_values ` method in `environment_loader.py` ([link](https://github.com/AI-for-Infrastructure-Management/imp-rl-challenge/blob/18a691043164264e90b283202794e73500b775a1/imp_rl_challenge/environments/config/environment_loader.py#L169)) to safeguard from such common errors.

> Rewards Q1: I would say if we halve the segment length, the speed and costs would double; therefore, it seems reasonable that there is a linear correlation between the two.

Actually, the cost wouldn't change. It's the same amount of work hours so you just contract a big group at once rather than a smaller group for a longer time?

> Rewards Q2: Normalizing rewards should be done by participants if needed IMO.

ok, let's leave it to them.

> Objective: I think we could include a bit more than just wage of the drivers, the cost of delayed goods, missed opportunity cost of next shipments, and worse utilization of the truck are all more costly if it takes more time to deliver I would say, but I am not sure how to arrive at a concrete number with this.

It is definitely interesting to get into those, but not easy to find data. Let finalise this env, we can tweak this value/factor later?

> 80 km/h base travel speed seems a bit slow. Did we not have 100 in mind?

We can do that too. Reduced speed is then 80kmph. The base travel time model is now

|              	| new speed 	| # days 	|  b'/b  	|
|--------------	|:---------:	|:------:	|:------:	|
| Do-nothing   	|    100    	|    0   	| 1.0000 	|
| Inspect      	|    100    	|    0   	| 1.0000 	|
| Minor-repair 	|     80    	| 28.125 	| 1.0157 	|
| Major-repair 	|     80    	| 106.25 	| 1.0618 	|
| Replacement  	|     80    	|   365  	| 1.2500 	|

> The capacity calculation uses 100% utilization for 24h/day for 356 days. Seems a bit much, so we could probably half this I would say.

I agree. The FHWA formula is not very clear to me. We could stick to the current value of 9M vehicles/year (~1k vehicles/hour), since we assume the same value for all segments.

**Reply by DanielHettegger** [reply]
>> Rewards Q1: I would say if we halve the segment length, the speed and costs would double; therefore, it seems reasonable that there is a linear correlation between the two.
>
> Actually, the cost wouldn't change. It's the same amount of work hours so you just contract a big group at once rather than a smaller group for a longer time?

Ah, yes you are right! Got it mixed up there for a second. In that case, the cost of replacement should not change even if we half the duration, right? (I guess we could even hire infinite teams and get it done immediately 😅)

**Comment by zieadm** [comment]
> I would really prefer if the replace action would not take more than one time step (We can double the cost and half the time as discussed). Because I feel like this makes the problem less nice to work with from a participant's view. But if we want to keep it, we need to have it observable to the agent in some way that the next action will be overridden.

I may say that I strongly agree with this point. This modification can also trigger some instabilities in early training phases, I have tried this before. Also, thinking on a 50-year horizon and following the same rationale behind the discount factor (i.e., future actions cost less), we can still assume that reconstruction in one year is, and will be more feasible throughout the planning scope.
